config/src/test/scala/com/typesafe/config/impl/ConfigSubstitutionTest.scala

```diff
@@ -924,7 +924,7 @@ class ConfigSubstitutionTest extends TestUtils {
 
     @Test
     def substSelfReferenceIndirect() {
-        val obj = parseObject("""a=1, b=${a}, a=${b}""")
+        val obj = parseObject("""b=${a}, a=${b}""")
```

Comment:
you can have a key twice to override or merge with a previous value, yep. It's fundamental to the library. Things do get weird if you make a loop like this, but that's why there are tests validating the exact behavior of all kinds of weird cases.

Most likely the test passing on 7 and failing on 8 is a bug in our code where we rely on something that wasn't guaranteed, but there are other possibilities of course.